Modify utils.trace API and convert some gadgets to it.

eiffel-fl opened the pull request:
Hi.


Jose and I are working on adding CLI to `biolatency` and `seccomp` gadget.
To do so, we need to modify the `utils.trace` API.

For the moment, I added several functions to this API:
1. `CreateTrace()` which creates a trace according to the given parameter.
2. `SetTraceOperation()` which updates the operation of an existing trace.
3. `PrintTraceOutput()` which prints the output of a trace.
4. `DeleteTrace()` which deletes a trace created by CreateTrace().

I also updated the following gadgets to this new API:
1. `process-collector`
2. `socket-collector`
3. `dns`

Code was tested and seems fine:
```bash
francis@machine:~/Codes/kinvolk/inspektor-gadget$ kubectl create ns demo                                        francis/utils-trace-new-api % u=
namespace/demo created
francis@machine:~/Codes/kinvolk/inspektor-gadget$ kubectl run --restart=Never -n demo --image=nginx nginx-app --port=80                
pod/nginx-app created
francis@machine:~/Codes/kinvolk/inspektor-gadget$ kubectl wait --timeout=-1s -n test-socketcollector --for=condition=ready pod/nginx-app ; kubectl get pod -n test-socketcollector
pod/nginx-app condition met
NAME        READY   STATUS    RESTARTS   AGE
nginx-app   1/1     Running   0          62m
francis@machine:~/Codes/kinvolk/inspektor-gadget$ ./kubectl-gadget socket-collector -n demo                     francis/utils-trace-new-api % u=
NODE        NAMESPACE    POD          PROTOCOL    LOCAL         REMOTE       STATUS
minikube    demo         nginx-app    TCP         0.0.0.0:80    0.0.0.0:0    LISTEN
francis@machine:~/Codes/kinvolk/inspektor-gadget$ ./kubectl-gadget process-collector -n demo                    francis/utils-trace-new-api % u=
NAMESPACE    POD          CONTAINER    COMM     PID       
demo         nginx-app    nginx-app    nginx    100307    
demo         nginx-app    nginx-app    nginx    100364    
demo         nginx-app    nginx-app    nginx    100365 
francis@machine:~/Codes/kinvolk/inspektor-gadget$ ./kubectl-gadget dns -n demo
# Jump to other terminal.
francis@machine:~/Codes/kinvolk/inspektor-gadget$ kubectl -n demo run mypod -it --image=praqma/network-multitool -- /bin/sh 
If you don't see a command prompt, try pressing enter.
/ # nslookup microsoft.com
# Go back to first terminal.
POD                            TYPE      NAME
mypod                          OUTGOING  microsoft.com.demo.svc.cluster.local.
mypod                          OUTGOING  microsoft.com.svc.cluster.local.
mypod                          OUTGOING  microsoft.com.cluster.local.
mypod                          OUTGOING  microsoft.com.
mypod                          OUTGOING  microsoft.com.
```

I was almost able to test it in a two nodes environment:
```
$ kubectl get node
NAME                                STATUS   ROLES   AGE   VERSION
aks-agentpool-17075730-vmss000001   Ready    agent   56m   v1.20.9
aks-agentpool-17075730-vmss000002   Ready    agent   56m   v1.20.9
$ ./kubectl-gadget process-collector -n kube-system
Failed to run the gadget on all nodes: [failed to create BPF collection: program dump_task: task iterator not supported]Failed to run the gadget on all nodes: None of them succeeded
$ kubectl get trace -A
No resources found
# Traces seems to be deleted
```
With the above problem and to be sure everything was OK regarding trace, I decided to comment out trace deletion in `RunTraceAndPrint`, then I run again the above commands:
```
$ ./kubectl-gadget process-collector -n kube-system
Failed to run the gadget on all nodes: [failed to create BPF collection: program dump_task: task iterator not supported]Failed to run the gadget on all nodes: None of them succeeded
$ kubectl get trace -A
NAMESPACE   NAME                      AGE
gadget      process-collector-kqhz4   5s
gadget      process-collector-stpqt   5s
```
This time, traces are present, so my modifications are "multinodes proof".


Best regards.

alban commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -104,169 +134,304 @@ func deleteTraces(contextLogger *log.Entry, traceRestClient *restclient.RESTClie
-		FieldSelector: fields.Everything().String(),
+func reconstructParamsFromTrace(trace *gadgetv1alpha1.Trace) *CommonFlags {
+	return &CommonFlags{
+		OutputMode: trace.Spec.OutputMode,
```
`CommonFlags.OutputMode` seems to be for things like `utils.OutputModeJson`, it does not seem related to `Spec.OutputMode` (except the name).

marga-kinvolk commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -41,9 +43,37 @@ import (
 
+// TraceConfig is used to contain information used to create a trace.
+type TraceConfig struct {
+	// SubCommand is gadget name, e.g. socket-collector.
```
Why not "GadgetName" then?

alban reviewed (commented):
Thanks for working on this. Some comments.

marga-kinvolk commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -104,169 +134,304 @@ func deleteTraces(contextLogger *log.Entry, traceRestClient *restclient.RESTClie
 	client, err := k8sutil.NewClientsetFromConfigFlags(KubernetesConfigFlags)
 	if err != nil {
-		contextLogger.Fatalf("Error in creating setting up Kubernetes client: %q", err)
+		return fmt.Errorf("Error in creating setting up Kubernetes client: %w", err)
```
```suggestion
		return fmt.Errorf("Error setting up Kubernetes client: %w", err)
```

marga-kinvolk commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -104,169 +134,304 @@ func deleteTraces(contextLogger *log.Entry, traceRestClient *restclient.RESTClie
+
+	nodes, err := client.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("Error in listing nodes: %w", err)
```
```suggestion
		return fmt.Errorf("Error listing nodes: %w", err)
```

marga-kinvolk commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -104,169 +134,304 @@ func deleteTraces(contextLogger *log.Entry, traceRestClient *restclient.RESTClie
+
+	traceRestClient, err := getRestClient()
+	if err != nil {
+		return fmt.Errorf("Error while getting trace rest client: %w", err)
```
```suggestion
		return fmt.Errorf("Error setting up trace REST client: %w", err)
```

marga-kinvolk commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -104,169 +134,304 @@ func deleteTraces(contextLogger *log.Entry, traceRestClient *restclient.RESTClie
+func updateTraceOperation(trace *gadgetv1alpha1.Trace) error {
+	traceRestClient, err := getRestClient()
+	if err != nil {
+		return fmt.Errorf("Error while getting trace rest client: %w", err)
```
This error doesn't add information. The getRestClient function already says exactly that.

marga-kinvolk commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -104,169 +134,304 @@ func deleteTraces(contextLogger *log.Entry, traceRestClient *restclient.RESTClie
+
+	patchBytes, err := json.Marshal(patch)
+	if err != nil {
+		return fmt.Errorf("Error marshalling results: %w", err)
```
```suggestion
		return fmt.Errorf("error marshalling the operation annotation: %w", err)
```

marga-kinvolk commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -104,169 +134,304 @@ func deleteTraces(contextLogger *log.Entry, traceRestClient *restclient.RESTClie
+// If no trace corresponds to this ID, error is set.
+func getTraceListFromID(traceID string) (gadgetv1alpha1.TraceList, error) {
+	var listTracesOptions = metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("trace-id=%s", traceID),
```
```suggestion
		LabelSelector: fmt.Sprintf("%s=%s", TRACE_ID, traceID),
```

marga-kinvolk commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -104,169 +134,304 @@ func deleteTraces(contextLogger *log.Entry, traceRestClient *restclient.RESTClie
+		Into(&traces)
 	if err != nil {
-		contextLogger.Fatalf("Error while getting rest config: %s", err)
+		return gadgetv1alpha1.TraceList{}, fmt.Errorf("Error getting traces from traceID %q: %w", traceID, err)
```
```suggestion
		return traces, fmt.Errorf("error getting traces from traceID %q: %w", traceID, err)
```

marga-kinvolk commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -104,169 +134,304 @@ func deleteTraces(contextLogger *log.Entry, traceRestClient *restclient.RESTClie
 		if err != nil {
-			deleteTraces(nil, traceRestClient, traceID)
-			contextLogger.Fatalf("Error creating trace on node %s: %q", node.Name, err)
+			return err
```
Here, I think you should actually add context information to the error.
```suggestion
			return fmt.Errorf("error updating trace operation for %s: %w", traceID, err)
```

marga-kinvolk commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -104,169 +134,304 @@ func deleteTraces(contextLogger *log.Entry, traceRestClient *restclient.RESTClie
-		if params.Node != "" && node.Name != params.Node {
+	for _, trace := range traces.Items {
+		if trace.Status.State == "Completed" {
+			return errors.New("Cannot change operation of Completed trace!")
```
What's with the exclamation sign?

marga-kinvolk commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -104,169 +134,304 @@ func deleteTraces(contextLogger *log.Entry, traceRestClient *restclient.RESTClie
 		if err != nil {
-			deleteTraces(contextLogger, traceRestClient, traceID)
-			contextLogger.Fatalf("Error getting traces: %q", err)
+			return fmt.Errorf("Error getting traces: %w", err)
```
This error doesn't add anything on top of the error message already returned by getTraceListFromID.
```suggestion
			return err
```

marga-kinvolk commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -290,32 +455,158 @@ RetryLoop:
 		// warnings together with the actual output generated by other nodes.
 		if successNodeCount == 0 {
-			printTraceFeedback(contextLogger.Warningf, nodeWarnings)
+			printWarningFunction := func(format string, args ...interface{}) { fmt.Printf(format, args) }
```
I think warnings should also go to stderr, and it should be prefixed with WARNING?

marga-kinvolk commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -290,32 +455,158 @@ RetryLoop:
 
+	client, err := k8sutil.NewClientsetFromConfigFlags(KubernetesConfigFlags)
+	if err != nil {
+		return fmt.Errorf("Error in creating setting up Kubernetes client: %w", err)
```
```suggestion
		return fmt.Errorf("Error setting up Kubernetes client: %w", err)
```

mauriciovasquezbernal replied:
hey @eiffel-fl, I still see `Error in creating setting up Kubernetes client` there.

marga-kinvolk commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -290,32 +455,158 @@ RetryLoop:
-		customResultsDisplay(contextLogger, nodes, &results)
+		nodes, err := client.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			return fmt.Errorf("Error in listing nodes: %w", err)
```
```suggestion
			return fmt.Errorf("Error listing nodes: %w", err)
```

marga-kinvolk commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -290,32 +455,158 @@ RetryLoop:
+	return nil
+}
+
+// DeleteTrace deletes the traces whom ID is given as parameter using
```
Traces aren't people.
```suggestion
// DeleteTrace deletes the traces for the given traceID using
```

marga-kinvolk commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -290,32 +455,158 @@ RetryLoop:
+func DeleteTrace(traceID string) error {
+	traceRestClient, err := getRestClient()
+	if err != nil {
+		return fmt.Errorf("Error while getting trace rest client: %w", err)
```
```suggestion
		return err
```

marga-kinvolk commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -290,32 +455,158 @@ RetryLoop:
+	return nil
+}
+
+// getTraceListFromParameters returns traces associated with the given arguments.
```
Why Parameters in the function name, but arguments in the description?

marga-kinvolk commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -290,32 +455,158 @@ RetryLoop:
+
+	traceRestClient, err := getRestClient()
+	if err != nil {
+		return []gadgetv1alpha1.Trace{}, fmt.Errorf("Error while getting trace rest client: %w", err)
```
```suggestion
		return []gadgetv1alpha1.Trace{}, err
```

marga-kinvolk commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -290,32 +455,158 @@ RetryLoop:
+		Into(&traces)
+
+	if err != nil {
+		return []gadgetv1alpha1.Trace{}, fmt.Errorf("Problem while doing GET to REST client: %w", err)
```
```suggestion
		return []gadgetv1alpha1.Trace{}, fmt.Errorf("error getting traces with the given parameters: %s: %w", labels, err)
```

marga-kinvolk commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -290,32 +455,158 @@ RetryLoop:
+		return []gadgetv1alpha1.Trace{}, fmt.Errorf("Problem while doing GET to REST client: %w", err)
+	}
+	if len(traces.Items) == 0 {
+		return []gadgetv1alpha1.Trace{}, fmt.Errorf("No traces found with the given parameters: %w", err)
```
```suggestion
		return []gadgetv1alpha1.Trace{}, fmt.Errorf("no traces found with the given parameters: %s: %w", labels, err)
```

marga-kinvolk reviewed (commented):
Another round of reviews. Sorry for the nitpicking, but previous experience trying to debug has made very sensitive to having good errors that add context when needed and don't just repeat what the function called already said.

mauriciovasquezbernal commented:
Some comments. 

Socket-collector and process-collector are working fine for me. I hit some issues with dns as the json mode is not working.

```
# works fine
$ ./kubectl-gadget dns 
POD                            TYPE      NAME
mypod                          OUTGOING  www.microsoft.com.default.svc.cluster.local.
mypod                          OUTGOING  www.microsoft.com.default.svc.cluster.local.
mypod                          OUTGOING  www.microsoft.com.svc.cluster.local.
mypod                          OUTGOING  www.microsoft.com.svc.cluster.local.
mypod                          OUTGOING  www.microsoft.com.cluster.local.
mypod                          OUTGOING  www.microsoft.com.cluster.local.
mypod                          OUTGOING  www.microsoft.com.
mypod                          OUTGOING  www.microsoft.com.

# doesn't work (output is not in json)
$ ./kubectl-gadget dns -o json
mypod                          OUTGOING  www.microsoft.com.default.svc.cluster.local.
mypod                          OUTGOING  www.microsoft.com.default.svc.cluster.local.
mypod                          OUTGOING  www.microsoft.com.svc.cluster.local.
mypod                          OUTGOING  www.microsoft.com.svc.cluster.local.
mypod                          OUTGOING  www.microsoft.com.cluster.local.
mypod                          OUTGOING  www.microsoft.com.cluster.local.
mypod                          OUTGOING  www.microsoft.com.
mypod                          OUTGOING  www.microsoft.com.
```

eiffel-fl commented:
Good catch!
I normally corrected it!

mauriciovasquezbernal reviewed (commented):
Well. I again put my comment in the wrong box and all my other comments weren't published. Trying again.

mauriciovasquezbernal commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -290,32 +448,160 @@ RetryLoop:
+	if err != nil {
+		return []gadgetv1alpha1.Trace{}, err
+	}
+	if len(traces.Items) == 0 {
+		return []gadgetv1alpha1.Trace{}, fmt.Errorf("No traces found with the given parameters %s: %w", labels, err)
+	}
```
I don't think it's actually an error. Just return an empty list and the caller could handle it. (Otherwise it's very difficult for the caller to understand if there was an error or if there are not traces).

mauriciovasquezbernal commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -208,65 +213,213 @@ func GenericTraceCommand(
+
+	var traces gadgetv1alpha1.TraceList
+
+	err = traceRestClient.
+		Get().
+		Namespace("gadget").
+		Resource("traces").
+		VersionedParams(&listTracesOptions, scheme.ParameterCodec).
+		Do(context.TODO()).
+		Into(&traces)
+	if err != nil {
+		return traces, fmt.Errorf("Error getting traces from traceID %q: %w", traceID, err)
+	}
+
+	if len(traces.Items) == 0 {
+		return traces, fmt.Errorf("No traces found for traceID %q!", traceID)
+	}
+
+	return traces, nil
```
I think you can create a common helper to use with `getTraceListFromID` and `getTraceListFromParameters` to avoid duplicating that code. I think it could receive as parameter the label selector (a string).

blanquicet commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -208,65 +213,236 @@ func GenericTraceCommand(
+	return traces, nil
+}
+
+// SetTraceOperation set the operation of an existing trace.
+// If trace does not exist or Annotations does not contain GADGET_OPERATION key
+// an error is returned.
```
I think this comment needs to be updated. This function does not return an error if the `Annotations` map does not contain the `GADGET_OPERATION` key anymore, correct?

blanquicet commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -208,65 +213,236 @@ func GenericTraceCommand(
+		return err
+	}
+
+	for _, trace := range traces.Items {
+		err = updateTraceOperation(&trace, operation)
+		if err != nil {
+			return fmt.Errorf("error updating trace operation for %s: %w", traceID, err)
+		}
+	}
```
I don't know if it could really happen but, theoretically, we could potentially have some traces that were updated and some others that don't. How can the user recover from this situation?

mauriciovasquezbernal replied:
hmm, it's a difficult one. Perhaps it should not return the first error that it finds but instead try to continue updating the operation for the other traces. An error should be returned if the operation failed in any trace. (We could try to create an error that contains all the error messages together).

blanquicet replied:
Yes, if we decide to try anyway to set the annotation on the other traces, I think we should collect all the errors and report them. Otherwise, it will be pointless to continue iterating.

blanquicet commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -290,32 +466,159 @@ RetryLoop:
+// and DeleteTrace().
+// This function is thought to be used with "one-run" gadget, i.e. gadget
+// which runs a trace when it is created.
+func RunTraceAndPrintCustomOutput(config *TraceConfig, customResultsDisplay func(nodes *corev1.NodeList, results *gadgetv1alpha1.TraceList)) error {
```
I know it is not strictly related to this PR but I think we can take the opportunity to remove the `nodes *corev1.NodeList` argument from the `customResultsDisplay()` function because it is not used. What do you think?

cc @alban

blanquicet reviewed (commented):
This is really a great refactoring and improvement. Thanks for working on this!

I didn't finish the review, I will continue tomorrow and I will also test it. I hope there are no outdated comments because you pushed some changes while I was reviewing it 😅 .

mauriciovasquezbernal commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -208,65 +213,213 @@ func GenericTraceCommand(
+		}
+	}
+
+	/*
+	 * It is possible we change this value in the above loop, restore it to its
+	 * previous value.
+	 */
```
Yeah. I agree. It looks weird.

mauriciovasquezbernal commented on cmd/kubectl-gadget/collector.go:
```diff
@@ -119,7 +118,9 @@ func processCollectorCmdRun(cmd *cobra.Command, args []string) {
-				contextLogger.Fatalf("Error marshalling results: %s", err)
+				fmt.Fprintf(os.Stderr, "Error marshalling results: %s", err)
+
+				return
```
This should also call `os.Exit(1)` to keep the same behavior as before.

mauriciovasquezbernal commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -41,9 +43,46 @@ import (
+	// NodeName is the name of the node to trace as given by the '--node' option.
+	// If empty, all nodes within the namespace will be traced.
+	NodeName string
+	// NamespaceName is the name of namespace to trace as give by the '-n' option.
+	// If empty, all namespaces will be traced.
+	NamespaceName string
```
I think just `Namespace` works better.

mauriciovasquezbernal commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -208,65 +213,241 @@ func GenericTraceCommand(
+	return traces, nil
+}
+
+// SetTraceOperation set the operation of an existing trace.
```
```suggestion
// SetTraceOperation sets the operation of an existing trace.
```

blanquicet commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -41,9 +45,39 @@ import (
+	OutputMode string
+	// Verbose is used to print more information.
+	Verbose bool
+}
```
The problem of setting it from the caller is that every gadget needs to do it while with the previous code, it was computed inside the shared code (`utils/trace.go`). I think Francis was following the same idea of the [utils.CommonFlags](https://github.com/kinvolk/inspektor-gadget/blob/02abdae390abdfb2889e9d52e0b7b2ef76e72c2c/cmd/kubectl-gadget/utils/flags.go#L49-L77) which does not have the namespace field.

BTW, why don't we just add the full `utils.CommonFlags` data structure inside `TraceConfig` instead of adding each field separately? In the end, [the only field](https://github.com/kinvolk/inspektor-gadget/blob/3852ae762c8e12785fd2ab49043ac1f9f96a330f/cmd/kubectl-gadget/seccompadvisor.go#L80-L91) that the gadgets will not use is the CustomColumns but I think it is an acceptable price to pay to keep things simpler, don't you think?

mauriciovasquezbernal replied:
> The problem of setting it from the caller is that every gadget needs to do it while with the previous code

The caller is already setting other fields, so one more should not make a big difference. 

> BTW, why don't we just add the full utils.CommonFlags data structure inside TraceConfig instead of adding each field separately?

Yes, that's a very good idea. I think it's already done at the time I'm commenting.

blanquicet commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -109,6 +143,375 @@ func deleteTraces(contextLogger *log.Entry, traceRestClient *restclient.RESTClie
+
+// PrintTraceOutput is used to print trace output.
+// Trace state must be either "Started" or "Completed" to contain some events.
+func PrintTraceOutput(traceID string, config TraceConfig,
```
I really like this suggestion but I am not sure about the name of those two functions. I would like the name to really describe what the function does. Following my suggestions but, of course, I am open to others 🙂.

Firstly, `PrintTraceGenericOutput` is used to receive the output through a stream and we pass it a function that transforms every single JSON entry received, into the requested output mode (JSON, Columns or CustomColumns). Therefore, why don't we call it `PrintTraceOutputFromStream`? In the end, it makes only sense for the gadgets that support `Spec.OutputMode=Stream`.

On the other hand, `PrintTraceCustomOutput` needs the traces to be completed because it will print the output from the `Status.Output` of each trace. In fact, we pass it a function that iterates over a TraceList to extract the `Status.Output` of each of them. Therefore, I suggest calling it `PrintTraceOutputFromStatus`. BTW, I think it is not necessary to send the entire list of traces but just a slice with all the `Status.Output`, but that's something we can change later because it would be out of scope for this PR.

Similarly, I suggest changing `RunTraceAndPrintGenericOutput` and `RunTraceAndPrintCustomOutput` into `RunTraceAndPrintStream` and `RunTraceAndPrintStatusOutput`.

What do you think?

blanquicet commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -208,65 +193,240 @@ func GenericTraceCommand(
-			deleteTraces(nil, traceRestClient, traceID)
-			contextLogger.Fatalf("Error creating trace on node %s: %q", node.Name, err)
+			traceID, present := trace.ObjectMeta.Labels[TRACE_ID]
+			if present {
```
Is there exist a case where it is not present?

blanquicet commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -290,32 +450,160 @@ RetryLoop:
+}
+
+// PrintAllTraces prints all traces corresponding to the given config.CommonFlags.
+func PrintAllTraces(config *TraceConfig) error {
```
I am not sure I understand what is the purpose of giving the possibility to pass this config here. I expect that it would allow me to filter the traces I want to be displayed, e.g. if I just want to see the traces in the `demo` namespace and so on. However, from what I see, this function does not work like that. Instead, it will show you the traces that match **exactly** the filter you use. For instance, let's start the biolatency gadget to create a trace:

```
$ ./kubectl-gadget biolatency start --node worker
Successfully created trace with ID: "AvN9nL9H4Q7hkxuR"!
You will need to use this ID to report trace.
In case you lost it, you can list existing traces with 'list' subcommand
```

However, I was expecting that the following command would show me all the traces in all the nodes and all the namespaces but it does not: 
```
$ ./kubectl-gadget biolatency list -A
NAMESPACE    POD    CONTAINER    TRACEID
```

Same here:
```
$ ./kubectl-gadget biolatency list --node worker -A
NAMESPACE    POD    CONTAINER    TRACEID
```

I will be only able to see my trace if the filter is identical:
```
$ ./kubectl-gadget biolatency list --node worker
NAMESPACE    POD    CONTAINER    TRACEID
default                          AvN9nL9H4Q7hkxuR
```

In my opinion, we should just show all the biolatency traces and that's it. What do you think? 

cc @mauriciovasquezbernal

blanquicet commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -290,32 +450,160 @@ RetryLoop:
+
+	w := tabwriter.NewWriter(os.Stdout, 0, 0, 4, ' ', 0)
+
+	fmt.Fprintln(w, "NAMESPACE\tPOD\tCONTAINER\tTRACEID")
```
I think we should add the nodes where it is running.

blanquicet commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -290,32 +450,160 @@ RetryLoop:
+			return errors.New("Trace has no ID!")
+		}
+
+		fmt.Fprintf(w, "%v\t%v\t%v\t%v\n",
+			filter.Namespace, filter.Podname, filter.ContainerName, id)
```
It is important to remark into the description of the `list` subcommand that we will add to the different gadgets, that this is the filter that the traces are using to run the gadgets and not the trace's information. I don't know if it is just me but I got confused.

mauriciovasquezbernal commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -92,9 +131,9 @@ func printTraceFeedback(f func(format string, args ...interface{}), m map[string
 	}
 }
 
-func deleteTraces(contextLogger *log.Entry, traceRestClient *restclient.RESTClient, traceID string) {
+func deleteTraces(traceRestClient *restclient.RESTClient, traceID string) {
 	var listTracesOptions = metav1.ListOptions{
-		LabelSelector: fmt.Sprintf("trace-template-hash=%s", traceID),
+		LabelSelector: fmt.Sprintf("%s=%s", TRACE_ID, traceID),
```
hmm, perhaps we can  call it `GLOBAL_TRACE_ID` to make it clear that traces on different nodes can share this ID? :thinking:

mauriciovasquezbernal commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -290,32 +454,243 @@ RetryLoop:
+
+			// If an entry with this traceID already exists, we just update the node
+			// name by concatenating it to the string.
+			var builder strings.Builder
+
+			builder.WriteString(info.nodeName)
+			builder.WriteString(",")
+			builder.WriteString(node)
+
+			info.nodeName = builder.String()
```
```suggestion
			info.nodeName = info.nodeName + "," + node
```

Does it work fine too?

mauriciovasquezbernal commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -290,32 +454,243 @@ RetryLoop:
+}
+
+// This type is used to create a string containing labels to filter out trace.
+type labelFilter struct {
+	GadgetName string
+	NodeName string
+	Namespace string
+	PodName string
+	ContainerName string
+	OutputMode string
+}
```
Please apply `go fmt`. (Applies also to other locations on this file).

mauriciovasquezbernal commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -290,32 +454,243 @@ RetryLoop:
+	labels := ""
+	separator := ""
+
+	v := reflect.ValueOf(filter)
```
I don't think using reflect for this is the best approach. What about using a `map[string]string` instead of `labelFilter`. You could just loop through all the elements of the map to create the label string. 
The caller could use something like:

```
filter := map[string]string{
		"GadgetName": config.GadgetName,
		"NodeName":   config.CommonFlags.Node,
...
```

mauriciovasquezbernal commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -41,9 +44,28 @@ import (
+	// e.g. a seccomp profile.
+	TraceOutputMode string
+	// CommonFlags is used to hold parameters given on the command line interface.
+	CommonFlags CommonFlags
```
I think it can be a pointer.

mauriciovasquezbernal commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -290,32 +454,243 @@ RetryLoop:
+
+	var listTracesOptions = metav1.ListOptions{
+		LabelSelector: labelsFromFilter(filter),
+		FieldSelector: fields.Everything().String(),
```
I know it was already existing. Can we remove it as it defaults to everything?

mauriciovasquezbernal commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -208,65 +194,243 @@ func GenericTraceCommand(
+	for _, trace := range traces.Items {
+		localError := updateTraceOperation(&trace, operation)
+		if localError != nil {
+			err = fmt.Errorf("%w\nError updating trace operation for %s: %w", err, traceID, localError)
```
Please check this line because it has two `%w`. (Github Actions message below)

mauriciovasquezbernal reviewed (commented):
Some minor comments. It's looking very good now.

I'll do another pass next week.

blanquicet commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -208,65 +195,243 @@ func GenericTraceCommand(
-		for _, i := range results.Items {
+		for _, i := range traces.Items {
+			trace = &i
 			if i.Status.OperationError != "" {
```
I am implementing the biolatency gadget on top of this PR and I found that this `for` loop needs to be modified because it will not work for the multi-rounds gadgets. 

I think it is because now each gadget needs to specify what is the `Status.State` it is waiting for. Without the multi-rounds gadgets, we avoided that need by adding an [internal logic](https://github.com/kinvolk/inspektor-gadget/blob/353f9a2a91211a9032cfd142cb21b884d8009b71/cmd/kubectl-gadget/utils/trace.go#L439) but now it is more difficult. Consider the following situation:

We create a trace and set the `start` operation:
```
$ ./kubectl-gadget biolatency start --node worker
P0HCLetshkIzyMzI
```

If the operation successes, the `Status.State` will go in `Started`. Then, if we execute `./kubectl-gadget biolatency stop P0HCLetshkIzyMzI`, this `for` loop will consider that the output is ready if the `Status.State` is `Completed` **or** `Started`. And this is the problem because our biolatency trace is already in `Stated` so this loop will immediately go out and it will not actually wait for the output (which will be ready when the `Status.State` moves to `Completed`). 

One solution is that the caller must specify which `Status.State` it is waiting for. For instance, in my previous example, it will be `Completed`, while the stream gadgets will always wait for `Started`. However, I think that the function `printTraceOutput` is already difficult to follow because it contains the logic for both output modes (Stream and Status) thus I don't want to add more complexity. 

Instead, I propose to move each output mode logic to their respective callers `PrintTraceOutputFromStream` and `PrintTraceOutputFromStatus` and the common function `printTraceOutput` could be renamed as `waitForTraceUpdate` which, among others, receives the `Status.State` it must wait for. In the end, this common function will contain mainly this `for` loop ([link](https://github.com/kinvolk/inspektor-gadget/blob/353f9a2a91211a9032cfd142cb21b884d8009b71/cmd/kubectl-gadget/utils/trace.go#L416-L478)) and then each caller will call the `genericStreamsDisplay` or the `customResultsDisplay`.

What do you think?

blanquicet reviewed (commented):
As I wrote in the big comment I added 😅, I am implementing the biolatency gadget on top of this PR and I think the current logic to wait for the trace's output needs to be modified to work with the multi-rounds gadgets.

I don't know if we prefer to keep this PR like this and make another PR to support the multi-rounds gadgets. In the end, it would be an additional support. What do you think? 

cc @mauriciovasquezbernal @alban @marga-kinvolk

mauriciovasquezbernal commented:
> I don't know if we prefer to keep this PR like this and make another PR to support the multi-rounds gadgets. In the end, it would be an additional support. What do you think?

I think we should try to merge this ASAP. We can iterate over in other PRs if needed.

eiffel-fl commented:
> I think we should try to merge this ASAP. We can iterate over in other PRs if needed.

Normally, I already addressed this comment and added the `waitForTraceOutput` function.
There is nonetheless a TODO inside it (regarding cookie) which will be addressed later.

blanquicet commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -287,36 +434,228 @@ RetryLoop:
 		// warnings together with the actual output generated by other nodes.
 		if successNodeCount == 0 {
-			printTraceFeedback(contextLogger.Warningf, nodeWarnings)
+			printWarningFunction := func(format string, args ...interface{}) { fmt.Fprintf(os.Stderr, "[WARN] "+format+"\n", args...) }
```
I am not sure I like that prefix `[WARN/ERR]` because not all the logs have it:
```
$ ./kubectl-gadget process-collector -n non-exixting-namespace
[WARN] Failed to run the gadget on all nodes: No container matches the requested filter
Failed to run the gadget on all nodes: None of them succeeded
```
Isn't there a way to make what we were doing with the `contextLogger` but at `main` level so that, as you said, we avoid passing the `contextLogger` to every function? I am thinking of something like what we are doing with the `params utils.CommonFlags` which are accessible from everywhere ([link](https://github.com/kinvolk/inspektor-gadget/blob/4dc3ae6c642e195117a84204c863bef41848bc31/cmd/kubectl-gadget/main.go#L27)).

@mauriciovasquezbernal any suggestion?

mauriciovasquezbernal replied:
I think `printTraceFeedback` is too complex now. Why not just to print the messages without the prefix? (we can remove the callback parameter and so on...)
If we really want to print the prefix I think we should use a logger everywhere, otherwise we'll reimplementing the wheel.

blanquicet reviewed (commented):
I think it is great how the DNS, the socket and process collector CLIs look because even if we are completely changing the `utils.trace` API, they continue looking simple and clear. That's great!

I tested the latest changes and it is working correctly with multi-run gadgets! 

I added some more comments. After those ones, I think we will be ready 🎉 .

mauriciovasquezbernal reviewed (commented):
Some final minor comments. It's good to merge after those are fixed IMO.

mauriciovasquezbernal commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -325,6 +664,13 @@ func genericStreamsDisplay(
 
+	client, err := k8sutil.NewClientsetFromConfigFlags(KubernetesConfigFlags)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error setting up Kubernetes client: %w", err)
```
btw, the CI is failing in this line because "Fprintf call has error-wrapping directive %w".

blanquicet commented on cmd/kubectl-gadget/collector.go:
```diff
@@ -165,11 +167,22 @@ func processCollectorCmdRun(cmd *cobra.Command, args []string) {
+
+	err := utils.RunTraceAndPrintStatusOutput(config, callback)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "%s\n", err)
```
@mauriciovasquezbernal I agree we shouldn't return `0` but I am wondering what would be the difference between using `RunE` and returning `err` with respect to print `err` and call `os.Exit(1)`. If there is not, I suggest using `RunE`. What do you say?

mauriciovasquezbernal replied:
`RunE` prints the usage by default if an error happens (which is not right in this case). We can use `SilenceUsage` to avoid that but not sure if it's worthy now. I'd say that we'd need to make a general pass over all the code to handle all these cases in the case way after this PR is merged.

mauriciovasquezbernal commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -287,44 +430,231 @@ RetryLoop:
+			continue
+		}
+
+		// If labels is not empty, we will need to separate the other label value
+		// with a comma, so change separator value.
+		if labels != "" {
+			separator = ","
 		}
```
I think you can just do `separator = ","` unconditionally after L509.

blanquicet commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -287,44 +430,231 @@ RetryLoop:
+
+	defer DeleteTrace(traceID)
+
+	if config.TraceOutputMode != "Stream" {
+		return errors.New("TraceOutputMode must be Stream. Otherwise, call RunTraceAndPrintStatusOutput!")
+	}
```
What about doing this before creating the Trace? I think it is not necessary to create it if we are going to delete it because the `TraceOutputMode` is not compatible with this function. WDYT?

blanquicet commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -287,44 +430,231 @@ RetryLoop:
+
+	defer DeleteTrace(traceID)
+
+	if config.TraceOutputMode == "Stream" {
+		return errors.New("TraceOutputMode must not be Stream. Otherwise, call RunTraceAndPrintStream!")
+	}
```
ditto.

blanquicet reviewed (commented):
I will do another run of tests tomorrow but I think we are ready 🎊

mauriciovasquezbernal reviewed (approved):
I think it's good to go! 

@eiffel-fl thanks for handling all the comments! (please don't merge until @blanquicet gives it another pass).

blanquicet commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -208,77 +198,230 @@ func GenericTraceCommand(
 
-	if params.Node != "" && !nodeFound {
-		contextLogger.Fatalf("Invalid filter: Node %q does not exist", params.Node)
+	// It is possible we change this value in the above loop, restore it to its
+	// previous value.
+	trace.Spec.Node = traceNode
+
+	if trace.Spec.Node != "" && !nodeFound {
```
I think you can directly use `traceNode` in the `if` statement. Or what did I miss?

blanquicet commented on cmd/kubectl-gadget/utils/trace.go:
```diff
@@ -287,44 +430,227 @@ RetryLoop:
+	}
+
+	defer DeleteTrace(traceID)
+
```
Please remove the one empty line 🙂

blanquicet reviewed (approved):
I think it looks really good and it is ready to be merged! @eiffel-fl 👏🏽
